### ISSUE TITLE
RfC: Add batch renaming functionality to light table.

### ISSUE THREAD

**Issue by schwerdf** [darktable-org/darktable#2247]
I recently submitted a pull request (#2082) adding to the Lua API the capacity to rename images in-database, eliminating the need to remove and re-import.

I have now been requested to start a discussion on adding similar functionality to the Darktable GUI. This is a more complex task, because the Lua API functions operate over only one image, but the GUI functionality would operate over all selected images. Thus, adding this functionality essentially means implementing a batch renamer within Darktable.

This issue is meant to collect comments about the best way to do this.

I will start off the discussion with my own proposal:
* Add a new light table module, "batch rename," containing the following widgets:
  * A combo box with options "move" and "copy".
  * A text box, used to input a renaming pattern, supporting the same system of variables (`$(FILE_FOLDER)`, etc.) as the export path and naming-pattern preferences.
  * A "rename" button, to execute the renaming.
* To be consistent with the current image-move/copy functionality:
  * If the "ask before moving images from film roll folder" preference is set, a confirmation dialog will appear before renaming starts.
* To be consistent with the current export functionality:
  * If a directory specified by the renaming pattern does not exist and/or is not registered as a film roll, it will be duly created and/or registered.

One open question is how to handle the case when a file specified by the renaming pattern already exists. To be consistent with the current image-move functionality, the rename should just fail on such a file, but I can think of use cases where behavior akin to the "create unique filename" export option would be better.

**Comment by shiro** [comment]
Any activity on this? Sounds like a much needed feature.

IMO it should be possible to use separate variables for time values such as month, day, minute, etc. when the picture was created in order to support any naming conventions.

By default failing on existing filenames sounds like a good idea, however the option to generate unique names sounds great as well.

**Comment by schwerdf** [comment]
No. As the Darktable developers [emphatically disagree](https://www.darktable.org/about/faq/#faq-rename-files) about this being a needed feature, I did not want to attempt it without _some_ kind of feedback (and at the moment I am not familiar enough with the implementation of the variable system to know how feasible my proposed approach is).

There was also some erroneous behavior in the image copy function (which I just transliterated from Redmine as issue #4427) that I thought should be fixed first.

**Comment by quovadit** [comment]
Batch renaming would be absolutely pefect in darktable, here are some arguments:
- pro
  - when importing from a connected camera, there already is a batch-renaming functionality, with file naming pattern in preferences
  - with multiple devices (e.g. 2 memory cards or slr+smartphone) there is no chance to import using the consistent naming
  - I am forced to use a second tool for a fundamental photo-workflow-function
- con
  - "developers emphatically disagree about this being a needed feature"
    - Funny side-note: directly below the [faq-entry](https://www.darktable.org/about/faq/#faq-rename-files) about file manager, there still is the statement, that there is no need to change the pixel pipeline
    - competitors like capture one pro, lightroom and digiKam do have the bulk-rename feature, so it seems to be "needed"

**Comment by wpferguson** [comment]
I've written a script that will do this including variable substitution and there is already a transfer_hierarchy script to move things around.  @johnny-bit I saw this one while I was looking for the other one :smile:

**Comment by bbinet** [comment]
Also interested in this feature request.
@wpferguson Would you agree to share you script here?

**Comment by wpferguson** [comment]
The current version is attached to one of the comments here https://github.com/darktable-org/lua-scripts/issues/269.  I need to rewrite it because there are a couple of issues.  Once that's done I'll add it to the lua-scripts repository.

**Comment by bbinet** [comment]
Ok, thank you @wpferguson

**Comment by mlaverdiere** [comment]
This is a much needed feature in dt (not just as a lua script), for all the reasons already identified. 

I totally understand that a developer may not want to spend time/energy on this kind feature or that there may be some technical or conceptual issues related to how it can be well integrated in dt. Alas, I'm not a dev mysself, so I can't help...

But other than that, the other "philosophical" reasons sometimes invoked against the integration of "file management" capabilities in dt are somewhat hard to follow (like: "we don't want to touch/alter the file"). Currently, dt has already some half-baked file management capabilities: you can move, copy or delete (trash) files and you can find a moved (missing) directory (so indeed, dt can already "touch" the files, to the point of trashing them!). 

So basically, what’s missing in terms of file management is “only” to be able to rename files and directories and move directories within dt, in order to have the database and (if needed) the sidecars automatically updated (instead of asking users to use their file manager to do so and remove/reimport files, which is cumbersome and somewhat confusing). 

So please, keep going with your proposition!

**Comment by wpferguson** [comment]
> Alas, I'm not a dev myself, so I can't help...

> not just as a lua script

Then why are you specifying the implementation?

> So basically, what’s missing in terms of file management is “only” to be able to rename files and directories and move directories within dt, in order to have the database and (if needed) the sidecars automatically updated (instead of asking users to use their file manager to do so and remove/re-import files, which is cumbersome and somewhat confusing).

All of this functionality is currently implemented using lua scripts.

_Why implement these functions in a lua script?_

1. Speed.  The user can have a solution in hours or days, instead of months or years.
2. Solution tuning.  The solution can be "adjusted" to fit the needs of the community.  Functionality can be changed/added as necessary to address user requests/wants/desires.  Bad assumptions can also be corrected and the solution gets lots of testing.  This way if it gets moved to core then the solution is well understood, thus easier to implement.

**Comment by mlaverdiere** [comment]
> > Alas, I'm not a dev myself, so I can't help...
> 
> > not just as a lua script
> 
> Then why are you specifying the implementation?

Not being a developer doesn't mean that I can't have an opinion on what a regular user would expect, which is, IMHO, to have such functionalities available out of the box, without relying on some external "plugin" (lua script).

Now, if like you are kindly explaining, the strategy is to first experiment with a lua script before it gets integrated as a core functionality, I'm totally fine with that (for what it worth!).

If this discussion is reserved for developers, please let me know. I may be wrong, but to my understanding, this report was a call for discussion open to all (since there was no activity for some time and I have an interest in file rename functionality within dt, I thought it was appropriate to comment, instead of opening another "feature request").

**Comment by wpferguson** [comment]
> without relying on some external "plugin" (lua script)

Does this not solve the problem?  If so, please give an example of how it doesn't solve the problem so that we can better understand the problem.

Or, is this an opinion (i.e. I don't **like** red)?  

As a developer I can fix a problem, but I can't fix an opinion.

**Comment by mlaverdiere** [comment]
Ok, I'll try again: My basic point is just to suggest to that, someday, it would be nice to have the file rename feature available out of the box in dt, as it is something usually integrated in other DAM tools (Digikam, Lightroom, etc,) and users may be expecting that (see other reasons mentioned by previous commentators). This is just a suggestion and I'm not claiming any "right" to have the feature. As I already said, I understand that this an open source project, relying on gracious and precious dev time/efforts. 
 
Now, if you think that keeping the feature as a lua script is the way to go, I guess that's also an opinion and a quite legitimate one, that I respect. 

Nothing else to add in my case.

**Comment by johnny-bit** [comment]
Currently the import module has this feature via copy & import function.

So the only thing missing now is in-core renaming of files in collection via probably a third tab in selected images module...

**Comment by wpferguson** [comment]
I'm planning on adding the rename script to the repository soon (when 3.6 lands), so there will be an option.

**Comment by quovadit** [comment]
@wpferguson : did you have time to rewrite the lua-script? 

I tried to write a gui which uses dt_image_rename but it loops in dt_cache_get_with_caller (goto restart)

**Comment by wpferguson** [comment]
Yes, it's been part of the [lua-scripts repository](https://github.com/darktable-org/lua-scripts)  for awhile.  Sorry I forgot to update this

**Comment by quovadit** [comment]
that's absolutely awesome, thanks a lot !!! :tada:

**Comment by caigner** [comment]
The copy & import module let's you sort your images into base and sub directories with a pattern like this one:

$(JOBCODE)/$(EXIF_YEAR)/$(EXIF_YEAR)-$(EXIF_MONTH)-$(EXIF_DAY)/

I'd like to suggest adding an 'organize' feature to the selected images[s] module which allows the user to organize the selected images according to the same patterns and options. This should move images and corresponding sidecar files to the new location.

Alternatively, the rename_images lua script may be expanded to include
- jobcode
- base directory pattern
- sub directory pattern

and an 'ignore images without EXIF data' option
